File: firebase/openci/ts_functions/src/buildJob/cancelBuildJob.ts
import { logger } from "firebase-functions/v2";
import { HttpsError, onCall } from "firebase-functions/v2/https";

import { BuildJobStatus, getBuildJob, updateBuildJobStatus } from "../firestoreData";
import { verifyTeamMembership } from "../team/teamAuth";

interface CancelBuildJobRequest {
  buildJobId: string;
}

function requireNonEmptyString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new HttpsError("invalid-argument", `${field} is required`);
  }
  return value;
}

export const cancelBuildJob = onCall<CancelBuildJobRequest, Promise<{ success: true }>>(
  async (request) => {
    const auth = request.auth;
    if (!auth) {
      throw new HttpsError("unauthenticated", "Unauthenticated");
    }

    const buildJobId = requireNonEmptyString(request.data?.buildJobId, "buildJobId");
    const jobResult = await getBuildJob({ id: buildJobId });
    const jobData = jobResult.data.buildJob;
    if (!jobData) {
      throw new HttpsError("not-found", "Build job not found");
    }

    const currentStatus = jobData.status;
    if (currentStatus !== BuildJobStatus.QUEUED && currentStatus !== BuildJobStatus.IN_PROGRESS) {
      throw new HttpsError(
        "failed-precondition",
        `Cannot cancel a build job with status '${currentStatus}'`,
      );
    }

    const teamId = typeof jobData.teamId === "string" ? jobData.teamId : undefined;
    if (teamId) {
      await verifyTeamMembership(auth, teamId);
    }

    await updateBuildJobStatus({ id: buildJobId, status: BuildJobStatus.CANCELLED });
    logger.info("Build job cancelled", {
      buildJobId,
      callerUid: auth.uid,
      teamId,
      previousStatus: currentStatus,
    });

    return { success: true };
  },
);

File: firebase/functions_dart/bin/cancel_build_job.dart
import 'package:firebase_functions/firebase_functions.dart';
import 'package:openci_shared/openci_shared.dart';

class CancelBuildJobRequest {
  CancelBuildJobRequest({required this.buildJobId});

  final String buildJobId;

  factory CancelBuildJobRequest.fromJson(Map<String, dynamic> json) {
    final buildJobId = json['buildJobId'];
    if (buildJobId == null || buildJobId is! String || buildJobId.isEmpty) {
      throw InvalidArgumentError(
        'buildJobId is required and must be a non-empty string',
      );
    }
    return CancelBuildJobRequest(buildJobId: buildJobId);
  }
}

Future<Map<String, dynamic>> cancelBuildJob(
  CallableRequest<CancelBuildJobRequest> request,
  Firebase firebase,
) async {
  final auth = request.auth;
  if (auth == null) {
    throw UnauthenticatedError('Unauthenticated');
  }

  final buildJobId = request.data.buildJobId;
  final firestore = firebase.adminApp.firestore();

  final jobDoc = await firestore
      .collection(buildJobsCollection)
      .doc(buildJobId)
      .get();

  if (!jobDoc.exists) {
    throw NotFoundError('Build job not found');
  }

  final jobData = jobDoc.data()!;
  final job = BuildJob.fromJson(jobData);

  if (job.status != BuildJobStatus.QUEUED &&
      job.status != BuildJobStatus.IN_PROGRESS) {
    throw FailedPreconditionError(
      'Cannot cancel a build job with status \'${job.status.name}\'',
    );
  }

  final teamId = job.teamId;
  if (teamId == null || teamId.isEmpty) {
    throw FailedPreconditionError(
      'Build job is not associated with a team',
    );
  }

  final teamDoc = await firestore.collection(teamsCollection).doc(teamId).get();

  if (!teamDoc.exists) {
    throw NotFoundError('Team not found');
  }

  final teamData = teamDoc.data()!;
  final team = Team.fromJson(teamData);
  if (!team.members.contains(auth.uid)) {
    throw PermissionDeniedError(
      'You are not a member of this team',
    );
  }

  final now = DateTime.now().toUtc();
  final updatedJob = job.copyWith(
    status: BuildJobStatus.CANCELLED,
    completedAt: now,
    updatedAt: now,
  );
  await firestore
      .collection(buildJobsCollection)
      .doc(buildJobId)
      .set(updatedJob.toJson());

  logger.info('Build job $buildJobId cancelled by ${auth.uid} in team $teamId');

  return {'success': true};
}
